### DISCUSSION TITLE
How to set default value for a variable for every request and override occasionally?

### DISCUSSION THREAD

**Discussion by ncsibra** [AnWeber/vscode-httpyac#236] [Q&A] [answered]
Hi,

I'm trying to set a variable in the global script, and then use it in a response event, but sometimes I want to override it in the request scope.
The use case is, that I want to increment a global counter in the response hook for 7 requests out of 10 if the request is successful and the variable is true, but ignore the update for the rest of the 3 requests.
For me the logical way to do it is to set the variable to true in the global script and override only in the specific requests instead of specifying it before every request, this is reducing the noise.

Example code:
```ts
{{
    update='true';
    console.log(`global - update: ${update}`);
}}

{{@response
    console.log(`response - update: ${update}, code: ${response.statusCode}`);

    if (update == 'true' && response.statusCode == 200) {
        $global.counter+=1
    }
}}



###
// initial set
GET https://postman-echo.com/get

{{
    $global.counter=1
}}

###
// increment
GET https://postman-echo.com/get

###
// not increment
@update='false'
GET https://postman-echo.com/get
```

First I call the "initial set" request, this will set the counter.
If I call the "increment" request after that then the response event will work, increment the counter successfully.
But calling the "not increment" will ruin it and the global script does not reset the variable and I don't understand why.
Tried to use `exports.update` instead of just `update` in the scripts, but it was undefined in the response event.
The only way I was able to make it work was if I changed the "not increment" request to this:

```ts
{{
    update='false'
}}
GET https://postman-echo.com/get
```
I assume it's working because the global and the request script is using the same scope, but it's unnecessarily noisy just to set a variable.

What is the scope of the inline variable when set using the `@update='false'` syntax? And how can I access it from the global script?
Or what is the nicest and least noisy way to do this?

**Comment by AnWeber** [comment, marked as answer]
Sorry it took so long to reply. I seem to have accidentally clicked away the notification. 
The way via `update = 'true' is not intended and surprisingly works. Here the variable is set at the global object. But `@update = "false"` sets it in my variable context which wins in the following runs. The correct way would be to use `exports.update` but then you run into an error that you could work around.

- enable [`httpyac.useRegionScopedVariables`](https://github.com/AnWeber/vscode-httpyac/blob/main/package.json#L542) in vscode
- use `{{@request` to set `exports.update="true"`
- However, my preferred way would be to use the metadata here.
```
{{@response

    if (!httpRegion.metaData.noUpdate && response.statusCode == 200) {
        $global.counter+=1
    }
}}
###
// not increment
# @noUpdate
GET https://postman-echo.com/get
```

**Reply by ncsibra** [reply]
Hi Andreas,

No problem, thanks for the answer.
The metadata based solution works and it's nice and clean.

But, out of curiosity I tried the other solution, enabled `httpyac.useRegionScopedVariables`  in vscode then tested with this code:

```js
{{@request
    exports.update='true';
    console.log(`global - update: ${exports.update}`);
}}

{{@response
    console.log(`response - update: ${update}, exports.update: ${exports.update}`);

    if (update == 'true' && response.statusCode == 200) {
        $global.counter+=1
    }
}}



###
// initial set
GET https://postman-echo.com/get

{{
    $global.counter=1
}}

###
GET https://postman-echo.com/get

###
{{
    update='false'
}}
GET https://postman-echo.com/get

###
@update=false
GET https://postman-echo.com/get
```

Both "update=false" version returns `INFO: response - update: true, exports.update: undefined` in the console, so looks like it's not working or I misunderstood something.

**Reply by AnWeber** [reply]
@ncsibra An order problem. First the variable would be set and then overwritten with the script. Fail. Using `{{+` should work, but not tested (see [docs](https://httpyac.github.io/guide/scripting.html#events))

```
{{+
   exports.update='true';
}}
```

**Reply by ncsibra** [reply]
It still doesn't work the way it should, it looks like the global script is not always called at least it is not producing a log always and the counter is not incremented either.
Plus these scripts are already in the global region, so the `{{+` syntax shouldn't have any effect, at least what I understand from the documentation.

Anyway, I don't want to waste your time, so marked your original reply as the answer, thanks. :)

**Reply by AnWeber** [reply]
Yes, I would prefer the solution with metaData anyway. I would have to debug the other one myself. It wouldn't be worth it for me. thanks
